alchemist/alchemist-maps/src/main/java/it/unibo/alchemist/boundary/gpsload/impl/LoadGPSTraceGPX.java

```diff
@@ -0,0 +1,105 @@
+package it.unibo.alchemist.boundary.gpsload.impl;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.openstreetmap.osmosis.osmbinary.file.FileFormatException;
+
+import io.jenetics.jpx.GPX;
+import io.jenetics.jpx.Track;
+import io.jenetics.jpx.TrackSegment;
+import io.jenetics.jpx.WayPoint;
+import it.unibo.alchemist.boundary.gpsload.api.GPSFileLoader;
+import it.unibo.alchemist.model.implementations.positions.GPSPointImpl;
+import it.unibo.alchemist.model.implementations.routes.GPSTraceImpl;
+import it.unibo.alchemist.model.implementations.times.DoubleTime;
+import it.unibo.alchemist.model.interfaces.GPSPoint;
+import it.unibo.alchemist.model.interfaces.GPSTrace;
+
+/**
+ * Class that reads GPS tracks from gpx files. 
+ */
+public class LoadGPSTraceGPX implements GPSFileLoader {
+
+    private final Set<String> extensions;
+
+    /**
+     * 
+     */
+    public LoadGPSTraceGPX() {
+        extensions = new HashSet<>();
+        extensions.add("gpx");
+    }
+
+    @Override
+    public List<GPSTrace> readTrace(final URL url) throws FileFormatException, IOException {
+
+        final InputStream stream = url.openStream();
+        final List<GPSTrace> ret = getGPX(stream).tracks().map(this::getTrace).collect(Collectors.toList()); 
+        stream.close();
+        return ret;
+    }
+
+    @Override
+    public Collection<String> supportedExtensions() {
+        return extensions;
+    }
+
+    private GPX getGPX(final InputStream stream) throws FileFormatException {
+        Objects.requireNonNull(stream, "Input stream can't be null");
+        try {
+            return GPX.read(stream);
+        } catch (IOException e) {
+            throw new FileFormatException("the gpx file can't be read...sure is a gpx file valid?");
+        }
+    }
+
+    private GPSTrace getTrace(final Track track) {
+        /*
+         * check if track exist
+         */
+        Objects.requireNonNull(track, "request GPS track not found");
```

Comment:
There shouldn't be such checks in private methods. The caller should be responsible for these checks. If it's just internal consistency, assertions could be used. Let's think about it.